### PR TITLE
feat(cli): Dirk multi-node delegation support

### PR DESCRIPTION
**Pull request by merklefruit**
This PR adds support for creating signed delegations from multi-node Dirk instances (aka `DistributedAccount`).

Unfortunately, the BLS library we have used until now (`BLST`) doesn't support M-N threshold signature recovery.
See the related issue: https://github.com/supranational/blst/issues/123

Since this operation is not overly complex, I've decided to implement it manually in `recover_signature.rs`,
following the [C reference implementation in MCL](https://github.com/herumi/mcl/blob/328e26f45ba565d031f9570e68e3d61836a17d7c/include/mcl/lagrange.hpp#L16) directly for the Lagrange interpolation

The implementation is tested by generating a signed delegation against a 2-of-3 Dirk setup running locally and then verifying the signature against the combined public key.

One important caveat for testing is that the multi-node setup only works if you modify your `/etc/hosts` file to add aliases for localhost to `localhost-1`, `localhost-2`, `localhost-3` because it's the only way to have different valid TLS certificates on the same machine.

### CODE REVIEW

File: bolt-cli/src/commands/delegate/dirk.rs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    /// Test generating signed delegations using a remote Dirk signer (multi-node instance).
+    /// This test requires multiple instances of Dirk to be running.
+    ///
+    /// ```shell
+    /// cargo test --package bolt --bin bolt -- commands::delegate::tests::test_delegation_dirk_multi
+    /// --exact --show-output --ignored --nocapture
+    /// ```
+    #[tokio::test]
+    #[ignore = "Requires Dirk to be installed on the system"]
+    async fn test_delegation_dirk_multi() -> eyre::Result<()> {
```
Comment (merklefruit) [inline bolt-cli/src/commands/delegate/dirk.rs:182]: this is WIP.

File: bolt-cli/src/common/dirk/distributed.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+        for participant in &self.participants {
+            if signatures_with_ids.len() >= self.threshold {
+                break;
+            }
+
+            let url = participant_url(participant);
+            let mut conn = Dirk::connect(url.clone(), self.credentials.clone()).await?;
+
+            // TODO: might have to unlock the account here
```
Comment (thedevbirb) [inline bolt-cli/src/common/dirk/distributed.rs]: Is this stale?

**Review by thedevbirb — APPROVED** [review]
Great job. BLS is so cool

File: bolt-cli/src/commands/delegate/dirk.rs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+    /// Test generating signed delegations using a remote Dirk signer (single instance).
+    ///
+    /// ```shell
+    /// cargo test --package bolt --bin bolt -- commands::delegate::tests::test_delegation_dirk_single
+    /// --exact --show-output --ignored --nocapture
+    /// ```
+    #[tokio::test]
+    #[ignore = "Requires Dirk to be installed on the system"]
+    async fn test_delegation_dirk_single() -> eyre::Result<()> {
```
Comment (mempirate) [inline bolt-cli/src/commands/delegate/dirk.rs:145]: Very nice tests

File: bolt-cli/src/common/dirk/distributed.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+            // Every remote signer must sign the same message
+            let signature = match conn.request_signature(account_name.clone(), hash, domain).await {
+                Ok(signature) => signature,
+                Err(err) => {
+                    warn!(?err, "Failed to sign message with remote signer at {}", url);
+                    continue;
+                }
+            };
```
Comment (mempirate) [inline bolt-cli/src/common/dirk/distributed.rs]: Can we do this concurrently by pushing them into a `FuturesUnordered` or something similar for each participant?

File: bolt-cli/src/common/dirk/test_util.rs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+/// Returns the DIRK client URL and credentials, and the corresponding server process handle.
+pub async fn start_single_dirk_test_server() -> eyre::Result<(String, TlsCredentials, Child)> {
+    try_init_tls_provider();
+
+    // Check if dirk is installed (in $PATH)
+    if Command::new("dirk").arg("--help").status().is_err() {
+        bail!("DIRK is not installed in $PATH");
+    }
+
+    let test_data_dir = env!("CARGO_MANIFEST_DIR").to_string() + "/test_data/dirk_single";
```
Comment (mempirate) [inline bolt-cli/src/common/dirk/test_util.rs]: Nit but can use `concat!(env!("CARGO_MANIFEST_DIR"), "/test_data/dirk_single");`

**Review by mempirate — CHANGES_REQUESTED** [review]

File: bolt-cli/src/common/dirk/distributed.rs (hunk 2 above)
Comment (merklefruit) [reply bolt-cli/src/common/dirk/distributed.rs]: Yes, removing

File: bolt-cli/src/common/dirk/test_util.rs (hunk 5 above)
Comment (merklefruit) [reply bolt-cli/src/common/dirk/test_util.rs]: nice, done

File: bolt-cli/src/common/dirk/distributed.rs (hunk 4 above)
Comment (merklefruit) [reply bolt-cli/src/common/dirk/distributed.rs]: Yes but it would need to be throttled to avoid DDoSing dirk

Comment (merklefruit) [reply bolt-cli/src/common/dirk/distributed.rs]: done in the latest commit

**Review by mempirate — APPROVED** [review]
LGTM! Great work

**Pull request merged**